### PR TITLE
Slot category cannot be changed if its in use

### PR DESCRIPTION
**Pull request by abdullokhtolibjonov**

### CODE REVIEW

File: Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs [hunk 1]
```diff
@@ -17,6 +17,7 @@ public ParkingSlotEditVM(Models.ParkingSlot slot)
             this.IsAvailableForBooking = slot.IsAvailableForBooking;
             this.Category = slot.Category;
             this.ParkingZoneName = slot.ParkingZone.Name;
+            this.HasAnyActiveReservation = slot.Reservations.Any(slot => slot.IsActive == true);
```
Comment (DilshodMQ) [inline Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs]: not slot, reservation in lambda expression

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 2]
```diff
@@ -111,7 +111,11 @@ public IActionResult Edit(Guid id, ParkingSlotEditVM parkingSlotEditVM)
             if (ModelState.IsValid)
             {
                 var existingSlot = _slotService.GetById(parkingSlotEditVM.Id);
-
+                if (parkingSlotEditVM.HasAnyActiveReservation == true && existingSlot.Category != parkingSlotEditVM.Category)
+                {
+                    ModelState.AddModelError("Category", "Category cannot be changed it is in use");
```
Comment (DilshodMQ) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs]: you should check ModelState after adding ModelError

**Review by DilshodMQ — CHANGES_REQUESTED** [review]

File: Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs [hunk 3]
```diff
@@ -17,6 +17,7 @@ public ParkingSlotEditVM(Models.ParkingSlot slot)
             this.IsAvailableForBooking = slot.IsAvailableForBooking;
             this.Category = slot.Category;
             this.ParkingZoneName = slot.ParkingZone.Name;
+            this.HasAnyActiveReservation = slot.Reservations.Any(r => r.IsActive == true);
```
Comment (napster1991) [inline Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs]:
```suggestion
            this.HasAnyActiveReservation = slot.Reservations.Any(r => r.IsActive);
```

File: Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs [hunk 4]
```diff
@@ -28,6 +29,7 @@ public ParkingSlotEditVM(Models.ParkingSlot slot)
         [Required]
         public Guid ParkingZoneId { get; set; }
         public string? ParkingZoneName { get; set; }
+        public bool HasAnyActiveReservation { get; set; } = false;
```
Comment (napster1991) [inline Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs]:
```suggestion
        public bool HasAnyActiveReservation { get; set; };
```
Boolean property is false by default.

File: Parking-Zone/Areas/Admin/Views/ParkingSlot/Edit.cshtml [hunk 5]
```diff
@@ -26,7 +26,15 @@
             </div>
             <div class="form-group">
                 <label asp-for="Category" class="control-label"></label>
-                <select asp-for="Category" class="form-control" asp-items="Html.GetEnumSelectList<SlotCategoryEnum>()"></select>
+                @if (Model.HasAnyActiveReservation == true)
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Views/ParkingSlot/Edit.cshtml]:
```suggestion
                @if (Model.HasAnyActiveReservation)
```

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 6]
```diff
@@ -108,10 +108,16 @@ public IActionResult Edit(Guid id, ParkingSlotEditVM parkingSlotEditVM)
...
             }
 
-            if (ModelState.IsValid)
+            var existingSlot = _slotService.GetById(parkingSlotEditVM.Id);
+
+            if (parkingSlotEditVM.HasAnyActiveReservation == true && existingSlot.Category != parkingSlotEditVM.Category)
             {
-                var existingSlot = _slotService.GetById(parkingSlotEditVM.Id);
+                ModelState.AddModelError("Category", "Category cannot be changed it is in use");
+                return View(parkingSlotEditVM);
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs]: Remove this line.

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 7]
```diff
@@ -108,10 +108,16 @@ public IActionResult Edit(Guid id, ParkingSlotEditVM parkingSlotEditVM)
                 ModelState.AddModelError("Number", "This Slot number already exists");
             }
 
-            if (ModelState.IsValid)
+            var existingSlot = _slotService.GetById(parkingSlotEditVM.Id);
+
+            if (parkingSlotEditVM.HasAnyActiveReservation == true && existingSlot.Category != parkingSlotEditVM.Category)
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs]: 1. You cannot rely on property coming from view.
2. HasAnyActiveReservation is not used on view.

File: Tests/Controllers/ParkingSlotControllerTests.cs [hunk 8]
```diff
@@ -45,6 +45,14 @@ public ParkingSlotControllerTests()
...
             Category = SlotCategoryEnum.Business,
             IsAvailableForBooking = true,
+            Reservations = new List<Reservation>()
+            {
+                new Reservation()
+                {
+                    StartTime = DateTime.Parse("2024-01-27 18:00:00"),
+                    Duration = 4
+                }
+            }
```
Comment (napster1991) [inline Tests/Controllers/ParkingSlotControllerTests.cs]: Please, explain, how these lines help to test new code.

**Review by napster1991 — CHANGES_REQUESTED** [review]

File: Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs [hunk 9]
```diff
@@ -17,6 +17,7 @@ public ParkingSlotEditVM(Models.ParkingSlot slot)
             this.IsAvailableForBooking = slot.IsAvailableForBooking;
             this.Category = slot.Category;
             this.ParkingZoneName = slot.ParkingZone.Name;
+            this.HasAnyActiveReservation = slot.Reservations.Any(r => r.IsActive);
```
Comment (DilshodMQ) [inline Parking-Zone/ViewModels/ParkingSlot/ParkingSlotEditVM.cs]: Get this property from slot

File: Tests/Controllers/ParkingSlotControllerTests.cs [hunk 10]
```diff
@@ -45,6 +45,10 @@ public ParkingSlotControllerTests()
             ParkingZoneId = Guid.Parse("dd09a090-b0f6-4369-b24a-656843d227bc"),
             Category = SlotCategoryEnum.Business,
             IsAvailableForBooking = true,
+            Reservations = new List<Reservation>()
+            {
+                new Reservation()
```
Comment (DilshodMQ) [inline Tests/Controllers/ParkingSlotControllerTests.cs]: use new() this  instead new Reservation()

**Review by DilshodMQ — CHANGES_REQUESTED** [review]

File: Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 11]
```diff
@@ -110,19 +110,18 @@ public IActionResult Edit(Guid id, ParkingSlotEditVM parkingSlotEditVM)
 
             var existingSlot = _slotService.GetById(parkingSlotEditVM.Id);
 
-            if (parkingSlotEditVM.HasAnyActiveReservation == true && existingSlot.Category != parkingSlotEditVM.Category)
+            if (existingSlot == null)
+            {
+                return NotFound();
+            }
+
+            if (existingSlot.HasAnyActiveReservation == true && existingSlot.Category != parkingSlotEditVM.Category)
```
Comment (napster1991) [inline Parking-Zone/Areas/Admin/Controllers/ParkingSlotController.cs:118]:
```suggestion
            if (existingSlot.HasAnyActiveReservation && existingSlot.Category != parkingSlotEditVM.Category)
```

**Review by napster1991 — APPROVED** [review]

**Pull request merged**